Tarball

reidmv opened the pull request:
This PR implements:

* An R10K::Tarball base class
* An R10K::Module::Tarball class, providing a "tarball" module type
* An R10K::Environment::Tarball class, providing a "tarball" environment type

Example usage:

```
---
production:
  type: tarball
  source: https://example.com/puppet-code-env-v1.0.0.tar.gz
  version: 99a906c99c2f144de43f2ae500509a7474ed11c583fb623efa8e5b377a3157f0 # sha256digest
  modules:
    puppetlabs-peadm:
      type: forge
      version: 2.5.0
    reidmv-xampl:
      type: git
      source: https://github.com/reidmv/reidmv-xampl.git
      version: 62d07f2
    reidmv-mymodule:
      type: tarball
      source: https://example.com/reidmv-mymodule-1.0.0.tar.gz
      version: 6128ada158622cd90f8e1360fb7c2c3830a812d1ec26ddf0db7eb16d61b7293f # sha256digest
```

Later enhancements built on top of this might include Artifactory types, or S3 types, with options for real versions, depending on how the remote repository works.

Solves #567

reidmv committed:
Initial tarball class

This class will serve as a base class for creating tarball environment
and module types.

reidmv committed:
Update initializers

Trying to make sure the global cachedir option gets respected by tarball

reidmv committed:
Update r10k.gemspec minitar dependency

It's not just a development dependency; it's an r10k dependency too.
Really it always has been because it's a puppetforge dependency.

reidmv committed:
Implement tarball module type

This permits modules of type tarball to be used. Tarball modules will
be extracted with no modification into the modules target directory.
Tarballs should not be packed to include a top-level containing folder.

    modules:
      lwf-remote_file:
        type: tarball
        source: https://forge.puppet.com/v3/files/lwf-remote_file-1.1.3.tar.gz
        version: bce01b849631926b452900eff5dc385c893c4320c95b290f992e3c529f417f0e

reidmv committed:
Implement environment tarball type

This permits core environment content to be sourced from tarballs. The
tarball environment type supports defining modules as well.

    ---
    production:
      type: tarball
      source: https://example.com/puppet-code-env-1.0.0.tar.gz
      version: 99a906c99c2f144de43f2ae500509a7474ed11c583fb623efa8e5b377a3157f0

reidmv committed:
Add basic test stubs for tarball objects

These aren't fully featured tests, but it at least gets the skeleton up
and structured.

reidmv committed:
Make tarball comply with delete-spec-dir option

reidmv committed:
Make tarball work to copy files on Windows

reidmv committed:
Add statically_defined_version for tarball module

I'm honestly not sure what this does, but it seems to be the cool new
thing; all the other modules have it.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+    def_setting_attr :cache_root, if R10K::Util::Platform.windows?
+                                    File.join(ENV['LOCALAPPDATA'], 'r10k', 'tarball')
+                                  else
+                                    File.expand_path(ENV['HOME'] ? '~/.r10k/tarball': '/root/.r10k/tarball')
```
Would it make sense to just do
```suggestion
                                    File.join(ENV['HOME'] || '/root', '.r10k', 'tarball')
```
?

mwaggett replied:
Do we need to account for `ENV['LOCALAPPDATA']` being unset?

mwaggett replied:
Also, is there a reason to have a different `cache_root` from other module types? I'm looking at https://github.com/puppetlabs/r10k/blob/main/lib/r10k/forge/module_release.rb#L16 and https://github.com/puppetlabs/r10k/blob/main/lib/r10k/forge/module_release.rb#L66 and wondering how much of that can be reused.

reidmv replied:
The user setting `cachedir` will dictate what is used here, just as it does [for git and forge cache roots](https://github.com/puppetlabs/r10k/blob/a722b71f49c8e8249f978ac443eb78afe877796f/lib/r10k/initializers.rb#L36-L38).

The value supplied is the default, and is [modeled after R10K::Git::Cache](https://github.com/puppetlabs/r10k/blob/a722b71f49c8e8249f978ac443eb78afe877796f/lib/r10k/git/cache.rb#L19-L29). That model does not account for `ENV['LOCALAPPDATA']` being unset.

It would be cleaner overall if there were a central way to define this default, but at present it doesn't seem like there is one. I think figuring out how to change that may best be done as a separate refactor PR.

reidmv replied:
Note: stumbled across settings.rb, and it looks like there may be a way to set defaults there. Probably worth reviewing. If so, then the default can be set there, and these classes (`R10K::Tarball`, `R10K::Git::Cache`) shouldn't need to set a default themselves... 🤔

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+
+    # @!attribute [r] sha256digest
+    #   @return [String] The tarball's expected sha256 digest
+    attr_accessor :sha256digest
```
Can we just call this `sha256`?

reidmv replied:
Honestly, I'm leaning towards `version` being the term that gets documented.

If we shorten it, I think `digest` (or the commonly used but slightly more technically ambiguous `checksum`) would be the preferred shortening. That's the part that conveys what this parameter is really about—the output of a hash function which can be used to validate the expected content.

I went with the very specific and technical `sha256digest` because I was trying to avoid the user needing to look up which algorithm to use for the `checksum`. However, we could probably expose that later by offering a `checksum_algorithm` if a use case arose for it.

I've put in a commit to change this to just `checksum` for now.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+    #        tarball content
+    #
+    # @return [Boolean]
+    def insync?(target_dir, purge: true)
```
Can you add a yardoc for the `purge` param? It's not super clear to me what it's being used for.

reidmv replied:
Renamed the parameter to `ignore_untracked_files` and added a yarddoc.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+    #        tarball content
+    #
+    # @return [Boolean]
+    def insync?(target_dir, purge: true)
```
```suggestion
    def in_sync?(target_dir, purge: true)
```
don't love smushing words together 😅

reidmv replied:
Yeah. I tend to agree, but `insync` as a term seems to be the r10k convention. As such, I think it's maybe best to leave it this way in service to the greater goal of consistency.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+    #
+    # @return [Boolean]
+    def insync?(target_dir, purge: true)
+      target_dir_entries = Find.find(target_dir).map(&:to_s) - [target_dir]
```
Looking at https://ruby-doc.org/stdlib-2.6.3/libdoc/find/rdoc/Find.html, I don't think this is giving you a collection of entries. You might just want `Dir[File.join(target_dir, '**', '*')]`?

reidmv replied:
I took a look into this, and while I'm not 100% sure why it works this way, it seems that `Find.find()` returns an iterator that will hit all the files it is supposed to, and if that iterator is passed to `#map`, it does what's expected. The end result is an array of strings.

The globbing behavior is actually not complete, as it misses dot-files. Given that `Find.find` is working, I think it's best to leave it. I suspect it's more reliable than globbing can be.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+        next if entry.directory?
+        return false unless file_digest(found) == reader_digest(entry)
+      end
+      purge ? target_dir_entries.empty? : true
```
Not really sure what `purge` is being used for, but it might make sense to put the `target_dir_entries.empty?` check higher up so that we can return early and not loop through tarball entries if there's nothing in the target dir.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+                      when %r{https?}
+                        download_to_file(src_uri, tempfile)
+                      else
+                        raise "Unexpected source scheme #{src_uri.scheme}"
```
This should probably raise an `R10K::Error`. And it would probably be helpful to include the `source` in the error message.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+        temp_digest = case src_uri.scheme
+                      when 'file', nil
+                        copy_to_file(src_uri.path, tempfile)
+                      when %r{[a-z]} # Windows drive letter
```
What does 'Windows drive letter' mean?

Also, this should be a much more restrictive regex - right now this will match literally anything with a lowercase letter in it (including 'https').

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+
+        # Verify the download
+        unless (sha256digest == temp_digest) || sha256digest.nil?
+          raise 'Downloaded file does not match digest'
```
This should also probably raise an `R10K::Error` and ideally include more info - what downloaded file? what does it not match?

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+        tempfile.binmode
+        src_uri = URI.parse(source)
+
+        temp_digest = case src_uri.scheme
```
Can we call this `temp_sha256` or something? `digest` is kinda general imo.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+                      end
+
+        # Verify the download
+        unless (sha256digest == temp_digest) || sha256digest.nil?
```
In what scenario would `sha256digest` be nil? That's something the user is supposed to define in their Puppetfile, right? Are we okay letting them not do that?

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+        begin
+          FileUtils.mv(tempfile.path, cache_path)
+        rescue Errno::EACCES
+          FileUtils.cp(tempfile.path, cache_path)
```
When would we expect to encounter this error? Is there also a case where we'd get this error, or another error, when attempting to `cp` instead of `mv`?

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+        end
+
+        tempfile.close
+        tempfile.unlink
```
Do these need to be in an `ensure`?

reidmv replied:
Yep, that would be preferred. Pushed a commit to address this by switching to `Tempfile::open` with a block.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+        File.exist?(cache_path) && sha256digest == file_digest(cache_path)
+      else
+        File.exist?(cache_path)
+      end
```
I would switch these around and use an `if` instead of `unless`. Using `unless` with `else` is hard to read imo.

mwaggett replied:
or you could just do
```suggestion
      return false unless File.exist?(cache_path)
      return false if sha256digest && sha256digest != file_digest(cache_path)
      return true
```

reidmv replied:
Good call. Adding a commit to simplify.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+    #
+    # @param reader [String] An object that responds to #read
+    # @return [String] The read data's sha256 hex digest
+    def reader_digest(reader)
```
Does it make sense to put these digest methods into some utility class that can be shared? They don't necessarily seem specific to this tarball feature.

Also I think they should include `sha256` in the names, since `digest` is fairly general.

reidmv replied:
The names were originally left generic in order to be applicable in the event the digest algorithm ever changes, or is made selectable per-tarball. I've tried making that more evident by using a `@checksum_algorithm` instance variable, and references such as the following. 

```
# At this time, the only checksum type supported is sha256. In the future,
# we may decide to support other algorithms if a use case arises. TBD.
@checksum_algorithm = :SHA256
```

```
def reader_digest(reader)
  digest = Digest(@checksum_algorithm)
```

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+      digest.hexdigest
+    end
+
+    def sanitized_dirname(string)
```
Isn't this defined elsewhere? We should probably pull it into a place that can be shared.

reidmv replied:
Created an `R10K::Util::Cacheable` module to share this method.

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+    end
+
+    def cache_sha256digest
+      raise "Cache not present" unless File.exist?(cache_path)
```
```suggestion
      raise R10K::Error, _("Cache not present at %{path}") % {path: cache_path} unless File.exist?(cache_path)
```

mwaggett commented on lib/r10k/tarball.rb:
```diff
@@ -0,0 +1,257 @@
+      string.gsub(/[^@\w\.-]/, '-')
+    end
+
+    def http_get(uri, method: Net::HTTP::Get, redirect_limit: 10, &block)
```
If you want this to be a generic http request method, I would change the method name to reflect that. Otherwise, if you want to keep it specific to GET, I would remove method as a param.

This also seems like something that should be shared, rather than specific to this class.

reidmv replied:
Removed `method:` as a user opt, and hard-coded to `Net::HTTP::Get`.

Regarding sharing, at the moment I don't know of anything else in R10k that needs to download files this way. I've added a comment to the method which clarifies exactly what it is doing:

```
# Start a Net::HTTP::Get connection, then yield the Net::HTTPSuccess object
# to the caller's block. Follow redirects if Net::HTTPRedirection responses
# are encountered, and honor settings[:proxy].
```

I think this could potentially be generalized for sharing, but I'm hesitant to break it off unless/until there's another consumer in r10k that wants to use it, for fear of over engineering. As near as I can tell R10K currently offloads all download-esque activity to other libraries. I'm considering `Net::HTTP` to be such a library, though, it's probably possible to use Faraday instead... I think we already have that as a dependency, if it gets us something `Net::HTTP` doesn't.

reidmv replied:
Reviewing Faraday, it looks like that library can do everything we need it to and more succinctly than `Net::HTTP`. Given that it's already an r10k dependency, the right thing to do may be to reduce this method to using Faraday instead.

reidmv replied:
I took a stab at using Faraday, but it looks like:

1. We want/need [streaming downloads](https://lostisland.github.io/faraday/usage/streaming), so I think we need a newer version of Faraday than r10k currently supports, due to the puppet-forge gem dependency constraints.
2. Using the newer version of Faraday, we run into a bit of a problem for testing; I can't figure out how to make Faraday's test stubs work with streaming downloads.

If these two problems didn't exist, it seems like we could fairly handily offload this to Faraday. I'm stopping that line of investigation for now though. Seems pointless given item 1.

If it did work, it _might_ look something like this.

```ruby
# @param uri [URI] The URI to download from
# @param output The file or path to save to
# @return [String] The downloaded file's sha256 hex digest
def download_to_file(uri, output)
  digest = DIGEST_CLASS.new

  File.open(output, 'wb') do |output_stream|
    faraday_connection.get(uri) do |req|
      req.options.on_data = Proc.new do |chunk, _|
        output_stream.write(chunk)
        digest.update(chunk)
      end
    end
  end

  digest.hexdigest
end

# Defined separately to enable test stubbing
def faraday_connection
  Faraday.new do |faraday|
    faraday.response(:follow_redirects)
    faraday.adapter(:net_http)
    faraday.proxy(settings[:proxy]) unless settings[:proxy].nil?
  end
end
```

reidmv replied:
Alright, last tweak for now: went ahead and moved all the file download/copy/checksum related methods out of `R10K::Tarball` and into a new utility module, `R10K::Util::Downloader`.

reidmv committed:
Rename tarball sha256digest to checksum

Use a more general, short term that indicates the *purpose* of the
value, without being tied to an implementation. This keeps config files
easier to read, and leaves the door open to changing the checksum
algorithm later, should the need arise.

reidmv committed:
Adjust Tarball#insync? param names for clarity

Add a yard doc for the optional parameter to R10K::Tarball#insync?, and
change the name to the more self-evident `ignore_untracked_files`.

reidmv committed:
Ensure tempfile is closed on error

Rather than using an explcit begin; ensure;, just use the Tempfile#open
block form.

Remove the enveloping tempdir wrapper. It's not clear what that is used
for.

reidmv committed:
Use if instead of unless; simplify cache_valid?

Per PR review suggestion

reidmv committed:
Centralize #sanitized_dirname method for DRY

Method used by both R10K::Git::Cache and R10K::Tarball

reidmv committed:
Improve error in lib/r10k/tarball.rb

Co-authored-by: Molly Waggett <mwaggett@alumni.reed.edu>

reidmv committed:
Remove opt, add comment to R10K::Tarball#http_get

We aren't doing anything besides Get, so just hard-code it for ease of
understanding.

The comment clarifies what this method does, and how it is used by
consuming methods.

reidmv committed:
Clarify checksum selectivity decision

R10K::Tarball only supports sha256 at this time, but the door is being
left open to potentially support other types of checksums in the future.

reidmv committed:
Strengthen regexes on Tarball scheme selector

reidmv committed:
Update lib/r10k/tarball.rb

Co-authored-by: Molly Waggett <mwaggett@alumni.reed.edu>

reidmv committed:
Centralize cachedir, proxy settings

Fix proxy and other setting propogation for R10K::Tarball. Use
R10K::Util::Cacheable to centrally default a default cachedir setting.
Can't set the default centrally yet because that might invalidate
existing caches, may need to do it on a major version change.

reidmv committed:
Cache dir and method name changes in R10K::Tarball

For consistency, clarity

reidmv committed:
Implmement R10K::Util::Downloader

Taking methods originally defined in R10K::Tarball and placing them into
a sharable utility module.

reidmv committed:
Make R10K::Tarball include R10K::Util::Downloader

This completes a move of generalizable methods from R10K::Tarball into
the sharable utility module R10K::Util::Downloader

reidmv committed:
Simplify cachedir key scheme

Just name cache files by either their checksum (if known), or download
URL (if no checksum given). While it's not gonna be super easy to
eyeball the cache and know what is or isn't there, it better guarantees
that the cache is fully efficient.

reidmv committed:
Allow tarball static version to use checksum

Previously, R10K::Module::Tarball#statically_defined_version only
accepted the :version key. However, :checksum also constitutes a
statically defined version, if it is passed.

Thought: #statically_defined_version should be called after #initialize.
Otherwise it's repeating work that's handled setting instance variables.

reidmv committed:
Implement test for R10K::Util::Downloader#http_get

This commit implements tests for R10K::Util::Downloader#http_get, and
fixes bugs found while doing that.
